### ISSUE TITLE
Testcard with Authentication Recured leaves order unpaid

### ISSUE THREAD

**Issue by kaspar-allenbach** [craftcms/commerce-stripe#96]
I'm testing a commerce instalation with a stripe testcard which required Authentication:

`4000 0027 6000 3184`

The payment redirects correctly and I'm greeted by the order success page of craft commerce.

But the order is marked unpaid in craft and in the stripe backend the order doesn't who up at all.

But when I use the test card without Authentication: `4242 4242 4242 4242`

Then all is nice and well. What do I miss? Is this due to the test environement?

**Comment by lukeholder** [comment]
Was the order's transaction marked as processing? It should be, and likely not marked as success as the webhook is not set up? Have you set up the webhook?

**Comment by AugustMiller** [comment]
Sorry to bump such an old thread, but I think we've been bit by this as well. Ours is a "legacy" installation using Tokens + Charges, but as far as I know still officially supported by Stripe (and capable of triggering the 3DS process).

When completing the 3D Secure redirection (and arriving back at `commerce/payments/complete-payment`), Commerce calls `ChargeResponse::setProcessing(true);`, which (as best I can tell) leads to [this](https://github.com/craftcms/commerce/blob/develop/src/services/Transactions.php#L455-L457) block evaluating, which marks the order as complete—despite the payment not being finalized/successful.

To be clear: _even when I explicitly fail the test-mode 3DS flow_, orders are completed upon reaching `complete-payment` due to the transaction being in a `processing` state.

If this is the intended behavior, I can understand needing to optimistically accept a payment that Stripe says is "Processing," but I would disagree that something known to have failed during the offsite/3DS flow should be treated the same way.

Initially, I thought this was the result of webhooks not firing properly (and therefore not updating transaction state), but that would suggest even more strongly that the Order should remain incomplete (i.e. if _no_ information is getting through, why would we assume it's completed?).

With webhooks configured on production and in a local environment (using the Stripe CLI), I'm getting mutex lock failures when the redirect and Webhook are served simultaneously:

```
Exception: Unable to acquire a lock for transaction: 3b7309b7796b8c347c3090b81712c197
```

After both requests fail (the webhook thread times out while waiting for the redirect to create the transaction, and the customer-facing `complete-payment` request fails because it can't get the lock), refreshing the failed page (hitting `complete-pament` again) completes the order (but doesn't capture the payment).

Not sure where to go, from here! I'm more than happy to share our application with the team over a screen share session. Been wrestling with this for about three days and truly can't get enough traction on it to decide what the root cause is (likely the lock issue), or what more I can do to investigate.

**Comment by mattstein** [comment]
@AugustMiller Is it possible to get this down to steps we can reproduce?

**Comment by lukeholder** [comment]
Some clear steps to reproduce the mutex lock error would help thanks.

As for the orders being completed yet "failed during the offsite/3DS flow":

> If this is the intended behavior, I can understand needing to optimistically accept a payment that Stripe says is "Processing," but I would disagree that something known to have failed during the offsite/3DS flow should be treated the same way.

We need to mark an order as complete otherwise the customer could continue to edit the cart, or make another payment. This is why we introduced the Payment Status. The order complete status by itself is not enough to know the order is ready for processing by staff.

You are right *if* we know that is failed at that point, we shouldn't complete it, but the async nature of the process makes it hard to know. I will see if we can improve it in any way - maybe there a way for us to get more information during the redirect back before the webhook comes in.

Reopening.

**Comment by AugustMiller** [comment]
Hey! Replying primarily to say THANK YOU for getting back so quick. I'll follow up with some more details in a moment.

Appreciate the paradigm correction, vis à vis 3DS success/fail not being synchronous—this helps contextualize the “early” call to `markAsComplete` in the payment lifecycle. Also suggests that our app is too eagerly providing some digital goods… so if nothing else, I've got some controllers to revise!

Back at you in a sec. 💞

**Comment by AugustMiller** [comment]
Ok, so check this out!

There are two places in Commerce that build identical mutex keys:

#### `craft\commerce\services\Payments::completePayment()`

https://github.com/craftcms/commerce/blob/develop/src/services/Payments.php#L394-L399

#### `craft\commerce\services\Webhooks::processWebhook()`

https://github.com/craftcms/commerce/blob/develop/src/services/Webhooks.php#L88-L95

This seems to suggest that concurrent requests to `complete-payment` and a webhook (not necessarily even to do with 3DS?) with the same transaction reference/hash would collide.

- If the front-end request arrives first, it would get the lock, write the transaction, and mark the Order as complete (good).
- If the webhook arrives first, Commerce immediately acquires a lock, then hands off processing to the Gateway. In our case, `craft\commerce\stripe\gateways\Gateway::handle3DSecureFlowEvent()` then enters a (max 30s wait) loop, waiting for the transaction to appear (via `complete-payment`). In that time, the other request may arrive—but it's held up by the lock key collision (max 15s wait), and never ends up writing the transaction data.

Hope this tracks. Still working on a full-fledged test case!

**Comment by AugustMiller** [comment]
As far as potential solutions go: would it be irresponsible to just `throw` instead `sleep`, and rely on Stripe retrying the webhook a little later?

Alternatively, just reducing the wait timer from 30s to ~5s (to get it clear of the `lock` wait) should be enough to yield back, right? Even in the current model, Commerce will reply with a 500-level status, which Stripe will interpret as it having failed.

**Comment by AugustMiller** [comment]
Sorry for the ping—just wanted to check if the above provided enough context to replicate the 3DS failure? I didn't have an opportunity to test on a totally fresh installation last week, but from what I can tell, there isn't anything fundamentally different about our implementation of the payment form.

One thing to note is that while I observe this in 100% of local and live checkouts, _it may actually depend on the user's geographic location, due to the relative latency between the client + server and Stripe and the server… or even Stripe's current webhook delivery performance!

Ultimately, I think we're looking at a “race condition”—whether that's academically/implicitly a bug or not, I haven't a clue—but it does seem to have left the door open to unpredictable performance.

Holler if you need anything else!

@lukeholder

**Comment by lukeholder** [comment]
@AugustMiller working on this today and will give you an update after my investigation

**Comment by lukeholder** [comment]
Closing this as fixed in Commerce 4.4 and Stripe 4.0
